### PR TITLE
Beta/split properties and references

### PR DESCRIPTION
**Pull request by tsmith023**
This PR marks a significant refactoring of the API surface to split `properties` and `references` out to become independent entities within the client. It does this so that references are more easily accessed both by non-generic and generic users

### CODE REVIEW

File: weaviate/collections/queries/base.py [hunk 1]
```diff
@@ -173,41 +173,35 @@ def __parse_nonref_properties_result(
 
     def __parse_ref_properties_result(
         self, properties: RepeatedCompositeFieldContainer[search_get_pb2.RefPropertiesResult]
-    ) -> dict:
+    ) -> Optional[dict]:
         if len(properties) == 0:
-            return {}
+            return None
```
Comment (dirkkul) [inline weaviate/collections/queries/base.py:183]: I think we should keep empty dicts

File: weaviate/collections/queries/base.py [hunk 2]
```diff
@@ -173,41 +173,35 @@ def __parse_nonref_properties_result(
...
-            properties=cast(T, self.__parse_result(props) if options.include_properties else {}),
+    ) -> _Object[Any, Any]:
+        return _Object(
+            properties=self.__parse_nonref_properties_result(props.non_ref_props)
+            if options.include_properties
+            else {},
             metadata=self.__extract_metadata_for_object(meta) if options.include_metadata else None,
+            references=self.__parse_ref_properties_result(props.ref_props)
+            if options.include_references
+            else None,
```
Comment (dirkkul) [inline weaviate/collections/queries/base.py:211]: same, I'd go for an empty dict if there are None

File: weaviate/collections/queries/base.py [hunk 3]
```diff
@@ -398,103 +413,103 @@ def _parse_media(media: Union[str, pathlib.Path, io.BufferedReader]) -> str:
...
-                    assert isinstance(existing_from_reference.return_properties, list)
-                    existing_from_reference.return_properties.append(nested_prop_name)
-            else:
-                if existing_from_reference.return_metadata is None:
-                    metadata = MetadataQuery()
-                else:
-                    metadata = existing_from_reference.return_metadata
-                setattr(metadata, nested_prop_name, True)
-                self.__from_references_by_prop_name[prop_name].return_metadata = metadata
+# class _PropertiesParser:
```
Comment (dirkkul) [inline weaviate/collections/queries/base.py:425]: remove if we do not need it?

File: weaviate/collections/queries/base.py [hunk 4]
```diff
@@ -251,123 +244,127 @@ def __result_to_group(
...
+        if not is_typeddict(generic_properties):
+            raise TypeError(
+                f"return_properties must only be a TypedDict or PROPERTIES within this context but is {type(generic_properties)}"
+            )
+        return _extract_properties_from_data_model(generic_properties)
 
-    def __parse_properties(self, return_properties: Optional[PROPERTIES]) -> Optional[PROPERTIES]:
-        return _PropertiesParser().parse(return_properties)
+    # def __parse_properties(self, return_properties: Optional[PROPERTIES]) -> Optional[PROPERTIES]:
+    #     return _PropertiesParser().parse(return_properties)
```
Comment (dirkkul) [inline weaviate/collections/queries/base.py:358]: also remove?

File: weaviate/collections/data.py [hunk 5]
```diff
@@ -305,11 +311,11 @@ def __init__(
...
-    def with_data_model(self, data_model: Type[TProperties]) -> "_DataCollection[TProperties]":
-        _check_data_model(data_model)
-        return _DataCollection[TProperties](
-            self._connection, self.name, self._consistency_level, self._tenant, data_model
-        )
+    # def with_data_model(self, data_model: Type[TProperties]) -> "_DataCollection[TProperties]":
+    #     _check_data_model(data_model)
+    #     return _DataCollection[TProperties](
+    #         self._connection, self.name, self._consistency_level, self._tenant, data_model
+    #     )
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/data.py:316]: Same as above

File: test/collection/test_queries.py [hunk 6]
```diff
@@ -1,75 +1,75 @@
...
-        (None, None),
-        ([None, "name"], ["name"]),
-        (["__article", "name"], ["name", FromReference(link_on="article")]),
-        ("__article", [FromReference(link_on="article")]),
-        ("__article__", [FromReference(link_on="article")]),
-    ],
-)
-def test_properties_parser_success(properties: Optional[PROPERTIES], output: Optional[PROPERTIES]):
-    assert output == _PropertiesParser().parse(properties)
+# @pytest.mark.skip(reason="string syntax has been temporarily removed from the API")
```
[the remark this answers is not in the document]
Comment (dirkkul) [reply test/collection/test_queries.py:13]: > Do you think it's better if we remove and then cherry-pick it later if we want it back?

In this case not, but a comment why it was commented out is really helpful - the test was already skipped so I wasn't sure why it was now also commented out

File: weaviate/collections/queries/base.py (hunk 3 above)
Comment (tsmith023) [reply weaviate/collections/queries/base.py:425]: Same as above

File: weaviate/collections/queries/base.py (hunk 4 above)
Comment (tsmith023) [reply weaviate/collections/queries/base.py:358]: Same as above

File: integration/test_collection.py [hunk 7]
```diff
@@ -123,19 +124,19 @@ class Right(TypedDict):
     assert isinstance(col, Collection)
 
 
-def test_data_with_data_model_with_dict_generic(client: weaviate.WeaviateClient):
-    name = "TestDataWithDictGeneric"
+# def test_data_with_data_model_with_dict_generic(client: weaviate.WeaviateClient):
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply integration/test_collection.py:127]: have added back :D

File: weaviate/collections/classes/types.py [hunk 8]
```diff
@@ -22,17 +25,38 @@
...
+        and not is_typeddict(properties)
+        # and not all([val in [
+        #     str,
+        #     int,
+        #     float,
+        #     bool,
+        #     list,
+        #     dict,
+        #     None,
+        # ] for val in get_type_hints(properties).values()])
```
[the remark this answers is not in the document]
Comment (tsmith023) [reply weaviate/collections/classes/types.py:57]: just gonna remove it for now, will add something similar in a smaller QoL PR

**Pull request merged**